ucp/useraccounts_module.php

```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
 						trigger_error($this->language->lang('FORM_INVALID') . '<br><br>' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 					}
 
-					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
+					// Query user badges
+					$sql = "SELECT badge_id
+						FROM $user_badge_table
+						WHERE user_id = $this_user_id";
+					if (!($result = $this->db->sql_query($sql)))
+					{
+						trigger_error($this->user->lang('UCP_UBADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+					}
+					$user_badges = $this->db->sql_fetchrowset($result);
+					$this->db->sql_freeresult($result);
 
 					// Delete all user badges that are NOT IN the array of marked badges or all if none is selected.
+					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
 					$sql = 'DELETE FROM ' . $user_badge_table . '
 						WHERE user_id = ' . $this_user_id . "
 						$delete_sql";
-					$this->db->sql_query($sql);
+					if (!$this->db->sql_query($sql))
+					{
+						trigger_error($this->user->lang('UCP_DELETE_UNSELECTED_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+					}
 
-					foreach ($markedbadges as $markedbadge)
+					if ($markedbadges)
 					{
-						// Reject duplicates.
-						$sql = "SELECT COUNT(*) as duplcount
-							FROM $user_badge_table
-							WHERE badge_id = $markedbadge
-							AND user_id = $this_user_id";
+						// Query valid badges
+						$sql = "SELECT badge_id
```

Comment:
On closer read, we can squash all three queries (including the `$user_badges` above, which is only used here) into a single query that lists all and only the valid badges:
```php
// Query valid badges
$sql_ary = array(
	'SELECT'	=> 'b.badge_id',

	'FROM'		=> array(
		$badge_table		=> 'b',
	),

	'LEFT_JOIN'	=> array(
		array(
			'FROM'	=> array($user_badge_table => 'bu'),
			'ON'	=> 'b.badge_id = bu.badge_id AND bu.user_id = ' . $this_user_id
		),
		array(
			'FROM'	=> array($badge_avail_table => 'ba'),
			'ON'	=> 'b.badge_id = ba.badge_id AND ba.user_id = ' . $this_user_id
		),
	),

	// Badges are valid if they are not already selected (bu.badge_id) and are either:
	//  - a default badge (b.badge_default)
	//  - have been awarded to the user (ba.user_id)
	'WHERE'		=> 'bu.badge_id IS NULL AND (b.badge_default = TRUE OR ba.user_id = ' . $this_user_id . ')',
);

$sql = $this->db->sql_build_query('SELECT', $sql_ary);
```

The check below then simplifies to `if ($this->core->in_2d_array($markedbadge, $valid_badges))`